### PR TITLE
Cache account settings

### LINKED ISSUE #13198: Cache account settings to improve performance
This task is to implement a cache to improve the performance when getting the account settings

### PR DESCRIPTION
**Pull request by atorresveiga**
Closes: #13198

### Description
This PR improves the account settings performance by saving the settings on the local preferences and displaying that information when available instead of only relying on the API request. With the changes introduced here, the local settings preferences will be the source of truth, and every time we fetch the account settings, we will update the preferences value.

### Testing information
TC1
1. Open the orders section
2. Open an order eligible for shipping label creation
3. Tap on create shipping label
4. Check that the account settings are fetched from the API
5. Navigate back
6. Tap again on create shipping label
7. Check that the account settings are presented from the preferences
8. Check that the account settings are refreshed on background

TC2
1. Open the orders section
2. Open an order eligible for shipping label creation
3. Tap on create shipping label
4. Check that the account settings are fetched from the API
5. Check that when the API request fails then the error screen is displayed

* You can apply the below path to force the fetch settings to fail
<details>
  <summary>Patch</summary>

```
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/FetchAccountSettings.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/FetchAccountSettings.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/FetchAccountSettings.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/FetchAccountSettings.kt	(revision afcf87f039221212a273e4c3a6b410a4424ce430)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/FetchAccountSettings.kt	(date 1735139698227)
@@ -10,19 +10,6 @@
     private val selectedSite: SelectedSite
 ) {
     suspend operator fun invoke(): Result<StoreOptionsModel> {
-        return selectedSite.getOrNull()?.let {
-            val response = shippingRepository.fetchAccountSettings(it)
-            val result = response.model
-            when {
-                response.isError.not() && result != null && result != StoreOptionsModel.EMPTY -> {
-                    Result.success(result)
-                }
-
-                else -> {
-                    val message = response.error?.message ?: "Unknown error"
-                    Result.failure(Exception(message))
-                }
-            }
-        } ?: Result.failure(Exception("No site selected"))
+        return Result.failure(Exception("No site selected"))
     }
 }

```

</details>

### The tests that have been performed
- Tested that the account settings are fetched from the preferences
- Checked that an API request is made to refresh the settings data
- Checked that when the request fails an error is displayed

### Images/gif

https://github.com/user-attachments/assets/754f7565-7877-4123-9189-0971569f63e1

- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>ShippingLabelsDataStore</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/datasource/ShippingLabelsDataStore.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import com.google.gson.Gson
+import com.woocommerce.android.datastore.DataStoreQualifier
+import com.woocommerce.android.datastore.DataStoreType
+import com.woocommerce.android.tools.SelectedSite
+import com.woocommerce.android.ui.orders.wooshippinglabels.models.StoreOptionsModel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+import javax.inject.Inject
+
+class ShippingLabelsDataStore @Inject constructor(
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/datasource/ShippingLabelsDataStore.kt:16]: There is nothing special happening here, so I decided to skip the unit tests for this class

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/WooShippingLabelRepository.kt [hunk 2]
```diff
@@ -29,6 +31,11 @@ class WooShippingLabelRepository @Inject constructor(
     ) = restClient.fetchAccountSettings(
         site = site,
     ).asWooResult { mapper(it.storeOptions) }
+        .also { response ->
+            response.model
+                ?.takeIf { response.isError.not() }
+                ?.let { dataStore.saveStoreOptions(it) }
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/WooShippingLabelRepository.kt:37]: When the response is a success, then persist the value in the preferences

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt [hunk 3]
```diff
@@ -105,415 +107,430 @@ class WooShippingLabelCreationViewModel @Inject constructor(
...
         launch {
-            fetchAccountSettings().fold(
-                onSuccess = {
-                    storeOptions.value = it
-                },
-                onFailure = {
-                    storeOptions.value = null
+            observeStoreOptions().collectIndexed { index, options ->
+                when {
+                    index == 0 && options == null -> {
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt]: We are using `null` as it is not found for account settings. If there is no value in the preferences, we will rely on the API request before updating the store options.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt [hunk 4]
```diff
@@ -105,415 +107,430 @@ class WooShippingLabelCreationViewModel @Inject constructor(
...
+                        refreshStoreOptions()
+                    }
+
+                    index == 0 && options != null -> {
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt]: We want to refresh the account setting only after consuming the first value

File: WooCommerce/src/main/kotlin/com/woocommerce/android/datastore/DataStoreModule.kt [hunk 5]
```diff
@@ -158,4 +159,24 @@ class DataStoreModule {
...
+    ) = PreferenceDataStoreFactory.create(
+        produceFile = {
+            appContext.preferencesDataStoreFile("shipping_label_configuration")
+        },
+        corruptionHandler = ReplaceFileCorruptionHandler {
+            crashLogging.recordEvent(
+                "Corrupted data store. DataStore Type: ${SHIPPING_LABEL_CONFIGURATION.name}"
+            )
+            emptyPreferences()
+        },
```
Comment (ThomazFB) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/datastore/DataStoreModule.kt:179]: Nice! Thanks for adding the corruption handler! 😄

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/ObserveStoreOptions.kt [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+import com.woocommerce.android.ui.orders.wooshippinglabels.datasource.ShippingLabelsDataStore
+import com.woocommerce.android.ui.orders.wooshippinglabels.models.StoreOptionsModel
+import kotlinx.coroutines.flow.Flow
+import javax.inject.Inject
+
+class ObserveStoreOptions @Inject constructor(
+    val dataStore: ShippingLabelsDataStore
+) {
+    operator fun invoke(): Flow<StoreOptionsModel?> = dataStore.observeStoreOptions()
+}
```
Comment (ThomazFB) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/ObserveStoreOptions.kt]: np: Do you think this use case is necessary, given it's serving only as an alias for a dataStore call? Or do you plan to expand the behavior later?

**Review by ThomazFB — APPROVED** [review]
Looks good! I left only a small regard over the new use case, but nothing critical or blocking.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/ObserveStoreOptions.kt (hunk 6 above)
Comment (atorresveiga) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/ObserveStoreOptions.kt]: Good point, Thomaz! I refactored this class to extract some logic from the model

Comment (ThomazFB) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/ObserveStoreOptions.kt]: Nice refactor, @atorresveiga! It makes sense to keep that DataStore data validation logic within the use case. Looks good to me!

**Pull request merged**